plugin/plugin_testconn.go

```diff
@@ -0,0 +1,70 @@
+package plugin
+
+import (
+	"context"
+	"fmt"
+)
+
+type TestConnFailureCode string
+
+const (
+	TestConnFailureCodeUnknown            TestConnFailureCode = "UNKNOWN"
+	TestConnFailureCodeUnimplemented      TestConnFailureCode = "UNIMPLEMENTED"
+	TestConnFailureCodeInvalidSpec        TestConnFailureCode = "INVALID_SPEC"
+	TestConnFailureCodeInvalidCredentials TestConnFailureCode = "INVALID_CREDENTIALS"
+)
+
+type TestConnError struct {
+	Code    TestConnFailureCode
+	Message error
+}
+
+func NewTestConnError(code TestConnFailureCode, err error) *TestConnError {
+	if code == "" {
+		code = TestConnFailureCodeUnknown
+	}
+	return &TestConnError{
+		Code:    code,
+		Message: err,
+	}
+}
+
+var ErrTestConnUnimplemented = &TestConnError{
+	Code: TestConnFailureCodeUnimplemented,
+}
+
+func (e *TestConnError) Error() string {
+	return e.Message.Error()
+}
+
+func (e *TestConnError) Unwrap() error {
+	return e.Message
+}
+
+func (e *TestConnError) Is(err error) bool {
+	if err2, ok := err.(*TestConnError); ok {
+		return e.Code == err2.Code
+	}
+	return false
+}
+
+// ConnectionTester is an interface that can be implemented by a plugin client to enable explicit connection testing.
+type ConnectionTester interface {
+	TestConnection(context.Context, []byte) *TestConnError
```

Comment:
Having this return `error` makes things weirder when we want to extract the failure code. Could be an interface that embeds `error`, and a `Code() string`...